### PR TITLE
Fixed 24782 -- Add TestCase.assertFormValid

### PR DESCRIPTION
**Pull request by delgiudices**
Before adding documentation and tests is this implementation okay?

### CODE REVIEW

**Comment by mjtamlyn** [comment]
Whilst `form.errors.as_json()` is passable, I'd love to see a more advanced implementation. Something like 
```
Form validation failed. The following field(s) had errors:
- username
    - This field is required
- password2
    - This field is required
    - Password fields must match
```

You may also like to position the code next to `assertFieldOutput` higher up the file as it belongs with the other form utility in tests.

File: django/test/testcases.py [hunk 1]
```diff
@@ -756,6 +756,18 @@ def assertXMLNotEqual(self, xml1, xml2, msg=None):
...
+    def assertFormValid(self, form, msg=None):
+
+        if not form.is_valid():
+            standardMsg = form.errors.as_json()
+            self.fail(self._formatMessage(msg, standardMsg))
+
+    def assertFormNotValid(self, form, msg=None):
+
+        if form.is_valid():
+            standardMsg = 'assertFormNotValid validated succesfully'
```
Comment (mjtamlyn) [inline django/test/testcases.py]: If we wish to implement the complement of this, here's some nicer pseudocode:
```
def assertFormNotValid(self, form, expected_errors=None, msg=None):
    to_report = []
    if form.is_valid():
        to_report.append('form was valid')
    else:
        for error in expected_errors:
            if not form.has_error(*error):
                to_report.append('missing %s' % error)
    if to_report:
        self.fail(to_report)
```

The idea is to optionally allow you to pass in a list of `(field, code)` tuples (this should be more flexible to the fact code is optional actually), which will be checked as having thrown errors. Extra errors are ignored (design decision?).

**Comment by mjtamlyn** [comment]
Calling @loic because I think he'll have design opinions.

**Comment by loic** [comment]
You can just iterate through `errors` itself to produce the output suggested by Marc.

```
In [1]: repr(f.errors)
Out[1]: '{\'charfield\': [u\'This field is required.\'], \'slugfield\': [u"Enter a valid \'slug\' consisting of letters, numbers, underscores or hyphens."]}'
```

There is also `as_data()` should you need even more info like the error code.
```
In [9]: repr(f.errors.as_data())
Out[9]: '{\'charfield\': [ValidationError([u\'This field is required.\'])], \'slugfield\': [ValidationError([u"Enter a valid \'slug\' consisting of letters, numbers, underscores or hyphens."])]}'
```

**Comment by delgiudices** [comment]
In what file exactly should i develop the tests for these methods? i was looking at django/tests/test_utils/tests.py but i don't think that's the right place

**Comment by timgraham** [comment]
I'd create a new file: `test_utils/test_assertformvalid.py`

File: django/test/testcases.py [hunk 2]
```diff
@@ -756,6 +756,36 @@ def assertXMLNotEqual(self, xml1, xml2, msg=None):
...
+        """
+        if not form.is_valid():
+            standardMsg = "Form validation failed. The following field(s) had errors:\n"
+            for field, field_errors in form.errors.items():
+                standardMsg += "- %s:\n" % field
+                for error in field_errors:
+                    standardMsg += "    - %s\n" % error
+            self.fail(self._formatMessage(msg, standardMsg))
+
+    def assertFormNotValid(self, form, expected_errors=None, msg=None):
```
Comment (timgraham) [inline django/test/testcases.py]: By the way, I don't find `assertFormNotValid` compelling as we already have `assertFormError`.

Comment (loic) [inline django/test/testcases.py]: True, unless we made `assertFormNotValid(expected_errors)` have to match the exact state of form errors. Currently both `assertFormNotValid` and `assertFormError` would pass if you only provide a subset of the errors that happened.

Not sure it's really worth it though.

Comment (mjtamlyn) [inline django/test/testcases.py]: I wasn't familiar with `assertFormError` actually. Guess the main difference is that `assertFormError` is a single field. Happy to leave that though, the nicer `assertFormValid` is useful in itself.

**Comment by mjtamlyn** [comment]
@loic and I discussed this today and decided there is a need for four separate functions here.

- `assertFormValid(form, msg=None)` is pretty much as implemented - it checks that a form is valid and if not nicely outputs the errors.
- `assertFormInvalid(form, expected_errors, allow_other_errors=True, msg=None)` is the complement of `assertFormValid`. It checks that the form is invalid, and asserts that a given set of fields have errors. Checking the exact message is not so supported here and should be discouraged as it is translatable/subject to copy changes. Checking the code associated with the `ValidationError` is supported and is encouraged. An option is provided to allow the set of errors passed to be non-exhaustive.
- `assertResponseFormValid(response,  expected_url, context_name='form', msg=None, **kwargs)` calls `assertRedirects(response, expected_url, **kwargs)` and if it fails (by returning a 200) will render similar output to `assertFormValid` failing. It takes an option `context_name` to find the relevant form in the 200 response context.
- `assertResponseFormInvalid(response, expected_errors, allow_other_errors=True, context_name='form', status_code=200, msg=None)` is an updated version of `assertFormError`, which should be deprecated. It is the complement of `assertResponseFormValid`, checking that the response does not redirect, and that the form in the context is invalid with a similar API to `assertFormInvalid`.

Notice that the first two are useful to unit testing forms, whereas the second two are used to simplify common tests for test client responses. `assertFormError` is deprecated because a) its API is restricted to checking a single field has an error and b) its naming doesn't suggest it is request specific.

Checking the code base in more detail, it seems there is also some justification in having formset related versions of these functions as well, and similarly deprecating `assertFormsetError`.

**Comment by timgraham** [comment]
To prevent needless code churn in existing projects, I'd suggest considering not deprecating the current methods.

**Comment by loic** [comment]
Let's ask @HonzaKral for his opinion on the formset methods.

**Comment by timgraham** [comment]
I committed the original implementation of `assertFormsetError()` after using it in my own projects via djangosnippets. I don't see a reason not to add analogous methods on the formset side.

**Comment by HonzaKral** [comment]
On formsets in `.as_data`we just need to be careful to return a list that corresponds to the list of forms. The forms that don't have any errors should imho be also present, just with empty dictionary.

**Comment by timgraham** [comment]
Please open a new PR if you continue work on the patch.

**Pull request closed without merging**